packages/custom-functions-metadata-plugin/src/customfunctionsplugin.ts

```diff
@@ -4,12 +4,16 @@
 import { generateCustomFunctionsMetadata, IGenerateResult } from "custom-functions-metadata";
 import * as fs from "fs";
 import * as path from "path";
-import { Compiler, sources, WebpackError } from "webpack";
+import { Compiler, sources, WebpackError, NormalModule } from "webpack";
 
 const pluginName = "CustomFunctionsMetadataPlugin";
 
 type Options = {input: string, output: string};
 
+declare global {
```

Comment:
This is to have a global map of results for each inputFile, right?  Does `declare global` need to be used? (I don't recall what that does off the top of my head.) I was thinking this could be a const object which is initialized to an empy object (rather than line 40 below) and then line 42 would put the results into the object.

Reply:
I though this would add the `generateResults` prop to the `typeof global` interface, but it looks like I was mistaken.  I'm going to change this into a declaration merging into the Global interface.

Reply:
I wasn't able to avoid using `declare global` (which I got from [the docs here](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#global-augmentation)). Maybe we could avoid it if we create a dedicated `.d.ts` file?

Reply:
I was thinking this could just be 

`const generateResults: { [key: string], IGenerateResult } = { };`

but I could be mistaken.

Reply:
We have to insert it on the `global` interface so we can store it like this:

```
global.generateResults[this.options.input] = generateResult;
```

So we can use it from the loader:
```
    global.generateResults[input].associate.forEach((item: IAssociate) => {
        source += `\nCustomFunctions.associate("${item.id}", ${item.functionName});`;
    });
```

Adding it to the global felt pretty hacky, but I just couldn't find a different way to share data between the plugin and the loader (which is required dynamically and its options are all `string | boolean | null`) without serialization.

Now given that this has to be a global, AFAIK the only way to patch the `Global` interface is either using `declare global` + interface merging or a custom `.d.ts` file.

Reply:
As an alternative, we could just make the loader build `generateResults` again. It would be done twice: at the plugin to generate the JSON metadata; at the loader to generate the associate calls.

I'd be fine with that too, but I don't know if that would be too much of a performance tax. Of even if a small performance penalty, if it's worth it to avoid using a global var.

Reply:
No, we don't want to do that. I'm still not clear why you can't just declare a const variable.

Reply:
If we declare a const variable inside `customfunctionsplugin.ts`, will the function inside `loader.ts` be able to access its value?

Reply:
Sorry I haven't had time to try that out. It was just a suggestion to see if that could be simplified, but I don't think it has to block the PR. If you're able to try it and get it to work, that would be great. If not, I think it's OK.

Reply:
Tried it, couldn't make it work

```
[0] src/customfunctionsplugin.ts(37,29): error TS2339: Property 'generateResults' does not exist on type 'Global'.      
[0] src/customfunctionsplugin.ts(38,28): error TS2339: Property 'generateResults' does not exist on type 'Global'.      
[0] src/customfunctionsplugin.ts(40,24): error TS2339: Property 'generateResults' does not exist on type 'Global'.      
[0] src/loader.ts(6,17): error TS2339: Property 'generateResults' does not exist on type 'Global'.
[0] src/loader.ts(6,44): error TS2339: Property 'generateResults' does not exist on type 'Global'.
[0] src/loader.ts(9,12): error TS2339: Property 'generateResults' does not exist on type 'Global'.
```

Reply:
If I just remove `global` references and try to get that variable from the scope, it also fails
```
[0] src/loader.ts(6,10): error TS2304: Cannot find name 'generateResults'.
[0] src/loader.ts(6,30): error TS2304: Cannot find name 'generateResults'.
[0] src/loader.ts(9,5): error TS2304: Cannot find name 'generateResults'.
```

Reply:
I have a change which compiles.

Reply:
![image](https://user-images.githubusercontent.com/741153/114757043-3c33d700-9d10-11eb-861e-6fd53164cb6a.png)
![image](https://user-images.githubusercontent.com/741153/114757095-49e95c80-9d10-11eb-817e-539545322ca1.png)

I could push a commit with this changes if they look OK to you.

Reply:
I didn't try it out.

Reply:
Ah, I see. You mean to export the variable. Because the plugin I used as base didn't do it, I assumed I shouldn't do it (it's obviously simpler). It's an hypothesis weakly held, though. I'll test it soon.

Reply:
We've got a situation here.
```
module.exports = CustomFunctionsMetadataPlugin;
```
This prevents us from exporting anything else. We can either change this into a named export (breaking change) or expose the `generateResults` as a static member.

Reply:
OK, I think you can make it a static member.